api/token.go

```diff
@@ -63,6 +65,50 @@ func (a *API) ResourceOwnerPasswordGrant(ctx context.Context, w http.ResponseWri
 	a.issueRefreshToken(user, w)
 }
 
+// AuthorizationCodeGrant implements the authorization_code grant for use with external OAuth providers
+func (a *API) AuthorizationCodeGrant(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	code := r.FormValue("code")
+	providerName := r.FormValue("provider")
+
+	if code == "" || providerName == "" {
+		sendJSON(w, 400, &OAuthError{Error: "invalid_request", Description: "Authorization code required"})
+		return
+	}
+
+	provider, err := a.Provider(providerName)
+	if err != nil {
+		BadRequestError(w, fmt.Sprintf("Unsupported provider: %+v", err))
+		return
+	}
+
+	tok, err := provider.GetOAuthToken(ctx, code)
+	if err != nil {
+		InternalServerError(w, fmt.Sprintf("Unable to exchange external code: %+v", err.Error()))
```

Comment:
I am on the fence here, so feel free to ignore, I don't think returning the error to the user is good. I'd rather have it be that we log the error in a detailed way and that we return a statement without the error.